rmf_site_editor/src/workspace.rs

```diff
@@ -264,208 +233,268 @@ pub fn dispatch_new_workspace_events(
+                .connect(scope.terminate)
+        });
+
+        let create_empty_workspace_from_dialog = world.spawn_workflow(|scope, builder| {
+            scope
+                .input
+                .chain(builder)
+                .map_async(|_| async move {
+                    #[cfg(not(target_arch = "wasm32"))]
+                    {
+                        if let Some(file) = AsyncFileDialog::new().save_file().await {
+                            let file = file.path().to_path_buf();
+                            let name = file
+                                .file_stem()
+                                .map(|s| s.to_str().map(|s| s.to_owned()))
+                                .flatten()
+                                .unwrap_or_else(|| "blank".to_owned());
+                            let data = WorkspaceData::LoadSite(LoadSite::blank_L1(
+                                name,
+                                Some(file.clone()),
+                            ));
+                            return Some(LoadWorkspaceFile(Some(file), data));
+                        }
+                        None
+                    }
+                    #[cfg(target_arch = "wasm32")]
+                    {
+                        let data =
+                            WorkspaceData::LoadSite(LoadSite::blank_L1("blank".to_owned(), None));
+                        Some(LoadWorkspaceFile(None, data))
+                    }
+                })
+                .then(process_load_files)
+                .connect(scope.terminate)
+        });
+
+        let load_workspace_from_path = world.spawn_workflow(|scope, builder| {
+            scope
+                .input
+                .chain(builder)
+                .map_async(|path| async move {
+                    let Some(data) = std::fs::read(&path)
+                        .ok()
+                        .and_then(|data| WorkspaceData::new(&path, data))
+                    else {
+                        warn!("Unable to read file [{path:?}] so it cannot be loaded");
+                        return None;
+                    };
+                    Some(LoadWorkspaceFile(Some(path.clone()), data))
+                })
+                .then(process_load_files)
+                .connect(scope.terminate)
+        });
+
+        let load_workspace_from_data = world.spawn_workflow(|scope, builder| {
+            scope
+                .input
+                .chain(builder)
+                .map_block(|data| Some(LoadWorkspaceFile(None, data)))
+                .then(process_load_files)
+                .connect(scope.terminate)
+        });
+
+        Self {
+            load_workspace_from_dialog,
+            create_empty_workspace_from_dialog,
+            load_workspace_from_path,
+            load_workspace_from_data,
+        }
+    }
+}
+
+impl WorkspaceLoadingServices {
+    /// Request to spawn a dialog and load a workspace
+    pub fn load_from_dialog(&self, commands: &mut Commands) {
```

Comment:
I don't have any objection to doing it this way, but I can suggest two alternatives that might be more ergonomic or idiomatic:

## 1. SystemParam

You could make a custom SystemParam struct that queries for both this resource and for `Commands`, then offers these `load_from_dialog`, etc methods and takes care of using Commands and the resource.

## 2. Extension trait for Commands

You could define a trait extension for Commands that implements these methods directly on Commands. However you'd also have to define one or more custom structs that implement the `Command` trait to get access to the resource from the `World` during the command flush, so this would be a bit more troublesome.

Reply:
+10 for the `SystemParam` approach, it's definitely a net gain so we can avoid this awkward `&mut commands` API, I forgot that it existed.
Added in https://github.com/open-rmf/rmf_site/pull/233/commits/5d08fbfb02681d3773ccd891b0667ac2146557a2

I do agree that the extension trait approach tends to be a bit boilerplate-y and harder to trace, but in any case we can mix and match and / or revisit when it's time to split the workcell editor mode